numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -1929,21 +1929,55 @@ array_asfortranarray(PyObject *NPY_UNUSED(ignored),
 
 
 static PyObject *
-array_copyto(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kwds)
+array_copyto(PyObject *NPY_UNUSED(ignored),
+        PyObject *const *args, Py_ssize_t len_args, PyObject *kwnames)
 {
-    static char *kwlist[] = {"dst", "src", "casting", "where", NULL};
-    PyObject *wheremask_in = NULL;
-    PyArrayObject *dst = NULL, *src = NULL, *wheremask = NULL;
+    PyObject *dst_obj, *src_obj, *wheremask_in = NULL;
+    PyArrayObject *src = NULL, *wheremask = NULL;
     NPY_CASTING casting = NPY_SAME_KIND_CASTING;
+    NPY_PREPARE_ARGPARSER;
 
-    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!O&|O&O:copyto", kwlist,
-                &PyArray_Type, &dst,
-                &PyArray_Converter, &src,
-                &PyArray_CastingConverter, &casting,
-                &wheremask_in)) {
+    if (npy_parse_arguments("copyto", args, len_args, kwnames,
+            "dst", NULL, &dst_obj,
+            "src", NULL, &src_obj,
+            "|casting", &PyArray_CastingConverter, &casting,
+            "|where", NULL, &wheremask_in,
+            NULL, NULL, NULL) < 0) {
         goto fail;
     }
 
+    if (!PyArray_Check(dst_obj)) {
+        PyErr_Format(PyExc_TypeError,
+            "copyto() argument 1 must be a numpy.ndarray, not %s",
+            Py_TYPE(dst_obj)->tp_name);
+    }
+    PyArrayObject *dst = (PyArrayObject *)dst_obj;
+
+    src = (PyArrayObject *)PyArray_FromAny(src_obj, NULL, 0, 0, 0, NULL);
+    if (src == NULL) {
+        goto fail;
+    }
+    PyArray_DTypeMeta *dtype = NPY_DTYPE(PyArray_DESCR(src));
+    Py_INCREF(dtype);
```

Comment:
Now that we vender pythoncapi-compat we can replace this with `Py_NewRef`. Not necessary here but we should think about doing a big refactoring codebase-wide because IMO it's way more readable.

Reply:
also can you maybe call this variable `DType`? Or maybe rename the descriptor in the if statement below to be `descr` instead of `dtype`? It's a little weird that the name dtype is bound to objects with two different types in this function

Reply:
Yeah, somehow didn't start yet using `Py_NewRef` even though I always wished it was there, definitely changing :).

Reply:
Will rename maybe both.  Actually, `Py_NewRef` isn't a macro so it's inconvinient if you are not working with `PyObject *`...

Reply:
I'd be happy having an `Npy_NewRef` that is the macro with a cast.